cli/commands/edit/edit.go

```diff
@@ -0,0 +1,109 @@
+package edit
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+
+	"github.com/sensu/sensu-go/cli"
+	"github.com/sensu/sensu-go/cli/commands/create"
+	"github.com/sensu/sensu-go/cli/commands/helpers"
+	"github.com/spf13/cobra"
+)
+
+// vi is the default editor!
+const defaultEditor = "vi"
+
+func extension(format string) string {
+	switch format {
+	case "yaml":
+		return "yaml"
+	case "wrapped-json", "json":
+		return "json"
+	default:
+		return "txt"
+	}
+}
+
+func Command(cli *cli.SensuCli) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "edit [RESOURCE TYPE] [KEY]...",
+		Short: "Edit resources interactively",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) < 2 {
+				_ = cmd.Help()
+				return errors.New("invalid argument(s) received")
+			}
+			resourceType := args[0]
+			resourceKeyParams := args[1:]
+			format := cli.Config.Format()
+			switch format {
+			case "yaml", "wrapped-json":
+			default:
+				format = "yaml"
+			}
+			ctlArgs := []string{
+				resourceType,
+				"info",
+			}
+			ctlArgs = append(ctlArgs, resourceKeyParams...)
+			ctlArgs = append(ctlArgs, "--format")
+			ctlArgs = append(ctlArgs, format)
+			originalBytes, err := exec.Command(os.Args[0], ctlArgs...).CombinedOutput()
+			if err != nil {
+				_, _ = cmd.OutOrStdout().Write(originalBytes)
+				return fmt.Errorf("couldn't get resource: %s", err)
+			}
+			tf, err := ioutil.TempFile("", fmt.Sprintf("sensu-resource.*.%s", extension(format)))
+			if err != nil {
+				return err
+			}
+			if _, err := tf.Write(originalBytes); err != nil {
+				return err
+			}
+			if err := tf.Close(); err != nil {
+				return err
+			}
+			editorEnv := os.Getenv("EDITOR")
+			if editorEnv == "" {
+				editorEnv = defaultEditor
+			}
+			execCmd := exec.Command(editorEnv, tf.Name())
+			execCmd.Stdin = os.Stdin
+			execCmd.Stdout = os.Stdout
+			execCmd.Stderr = os.Stderr
+			if err := execCmd.Run(); err != nil {
+				return err
+			}
+			changedBytes, err := ioutil.ReadFile(tf.Name())
+			if err != nil {
+				return err
+			}
+			if bytes.Equal(originalBytes, changedBytes) {
```

Comment:
I suppose that works and I doubt people will have huge objects, say with sizes on the order of dozens or hundreds of MB, that will make byte by byte comparison a problem.

Generally speaking though, I think your best bet is to use a hashing function over the content of the files (it doesn't have to be cryptographically strong here) and compare the resulting hashes.

Reply:
Individual resources will never even be a MB, we are good!